Update win32 global symbols

figment opened the pull request:
Update for win32 symbols based on my analysis.  Removed keydisplay since it crashes my df and I'm not familiar with the address.   

I am not familiar with testing and analysis tools used by the team so these have only been evaluated static code inspection between the 0.43.03 and 0.43.05 images.  

I offer the update in case it speeds up someone else's analysis.

lethosor commented on symbols.xml:
```diff
@@ -127,90 +127,90 @@
-        <global-address name='selection_rect'/>
-        <global-address name='gamemode'/>
-        <global-address name='gametype'/>
+        <global-address name='cursor'value='0x00f93040'/>
```
No spaces before `value`? I'm assuming DFHack doesn't choke on this, but it would be nice for readability purposes to have spaces.

lethosor commented on symbols.xml:
```diff
@@ -219,22 +219,22 @@
         generated standingorders
 
         <global-address name='standing_orders_mix_food' value='0x00f9303d'/>
-        <global-address name='standing_orders_gather_bodies' value='0x00f9303e'/>
```
Where did you get the new values for these? `devel/find-offsets standing_orders` should pretty much always be right, so I would be wary of static analysis that doesn't match that. (Of course, I don't know whether these are from find-offsets or not.)

figment replied:
I think 5 -6 of the ones in the original file were all the same address which is clearly wrong.  I'm not familiar with your code finding routines.

The standing_orders were found by locating the same 2-3 subroutines in both the 0.43.03 and 0.43.05 and comparing the order that the globals were referenced.  I then compared count of references to each address in other routines to verify.  Spot checked them by comparing routines from both versions where the code was used.

lethosor replied:
Yeah, they're definitely not all the same address. :)

You should still run `devel/find-offsets standing_orders`, though, just to verify that they're correct.

lethosor commented on symbols.xml:
```diff
@@ -127,90 +127,90 @@
         generated keydisplay
 
-        <global-address name='keydisplay' value='0x018b35fc'/>
+        <global-address name='keydisplay' />
```
Just to confirm, does
```
:lua ~dfhack.screen.getKeyDisplay(df.interface_key.SELECT)
```
crash for you?

figment replied:
Yes it does. in "string Screen::getKeyDisplay(df::interface_key key)" which is why I blanked it out.  As I cannot get through the initial menu for testing.   This is also after fixing the texture pointer so it wasn't that.

quietust replied:
I just did a bit of analysis (using the "update_keydisplay" function in g_src/enabler_input.cpp) and I've concluded that the actual location of keydisplay is 0x18B3600 - the old address was the very end of an std::string variable (namely, its allocated-length member).

Also keydisplay is not a vector - it is, and always has been, a std::map<InterfaceKey, std::set<std::string,less_sz> >;

figment replied:
Yes that address works.  The lua code above shows Enter for me.  And vector/map difference is noted, I noticed the code using begin() and end() and didn't pay more attention to it.

I'll try to find the Win64 and update my file.

lethosor commented:
There's no need to add leading `0`s to existing addresses. I thought they had changed entirely until I looked more closely.

I'm still curious about the standing_orders changes, though - how did you determine those?

figment commented:
Regarding the 0 prefix that is just my style for hex addresses and how I was tracking the addresses which was in a spreadsheet.

Also note that I updated the Win64 addresses. Swapped gamemode and gametype.  as well and fixed the Win32 texture address I had which I determined was wrong when checking the Win64 code of same.

Also fixed the whitespacing.  Sorry rookie mistake there.  Araxis Merge xml compare reformats the file so its hard to tell stuff like that.

lethosor commented on symbols.xml:
```diff
@@ -1220,8 +1224,8 @@
-        <global-address name='gamemode' value='0x140e60088'/>
-        <global-address name='gametype' value='0x140e6008c'/> -4
+        <global-address name='gamemode' value='0x140e6008c'/>
+        <global-address name='gametype' value='0x140e60088'/>
```
If these were found with find-offsets, as the comment above claims, I'm not sure how they could've been switched, since gamemode is 3 in the main menu and gametype is 11. Could you double-check those? It could be that @pronvit mixed them up, since one was corrected manually IIRC.

(Also, these are duplicated above, so you can probably get rid of the lines in this section.)

figment replied:
With my offsets I get the following on my Win64 build.
[lua]# ! df.global.gamemode
3
[lua]# ! df.global.gametype
11

The file had duplicates rows to start with. I'm not familiar enough with the scripts to know if that would cause issues.

Unrelated but 2 issues I have on the Win32 build that I dont have on the Win64 build with these pointers are:
1. The DFHack and Git text is truncated to 1 character.
2. The df generated seem to only be of 1 character width and shoved on left side. 

I assume these are same problem.  I can figure out how to join the irc channel if its better to discuss this there.  Not sure this is related to my offsets.

lethosor replied:
df::graphic is one we have the header for, so I'd be surprised if it's still incorrect. It definitely looks right in both builds on OS X and Linux.

pronvit replied:
So gametype/gamemode changed order now? Interesting.

pronvit replied:
@lethosor, need to check print_time field - array size is different in .h and .xml, and LARGE_INTEGER type it has is weird.

lethosor replied:
Yep, I mentioned that on IRC. The interesting thing is, I *think* `sizeof(LARGE_INTEGER)  == sizeof(long long) == sizeof(int64_t)`, so if "long" is always 32 bits on Windows, LARGE_INTEGER[100] should be the same size as long[200].

figment replied:
8 byte data structures will have 8 byte alignment boundaries I believe so it happens that Win32 build has 4 extra padding bytes before the print_time array that is not there on Win64. I verified that all is well if I write code that emits alignment decls and both my win32 and win64 work now. (also added explicit padding after display_frames and original_rect)
long rect_id;
__declspec(align(8)) long print_time[200];
long print_index;

I tried the following but had issues with getting both builds working in menus but should work in theory and would be a better solution. There may have been issues with the padding with other members.
long rect_id;
long long print_time[100];
long print_index;

figment commented:
8 byte data structures will have 8 byte alignment boundaries I believe so it happens that Win32 build has 4 extra padding bytes before the print_time array that is not there on Win64.  I verified that all is well if I write code that emits alignment decls and both my win32 and win64 work now.  (also added explicit padding after display_frames and original_rect)
    long rect_id;
    __declspec(align(8)) long print_time[200];
    long print_index;

I tried the following but had issues with getting both builds working in menus but should work in theory and would be a better solution.  There may have been issues with the padding with other members.
    long rect_id;
    long long print_time[100];
    long print_index;

lethosor commented:
Yes, but that's not the case in the actual DF headers: 
https://github.com/svenstaro/dwarf_fortress_unfuck/blob/2ba59c87bb63bea598825a73bdc896b0e041e2d5/g_src/graphics.h#L62-L65

```cpp
		long rect_id;

		LARGE_INTEGER print_time[100];
		long print_index;
```

I just added LARGE_INTEGER in bb582c8 - can you try updating your copy of df-structures and try again?

figment commented:
Build updated and it was working with the large_integer.  I still had explicit padding after display_frames and original_rect and cannot be bothered to rebuild and test yet again at this time. I did get my adventure mode region teleportation working on both Windows builds so I'm happy for now.
